### ISSUE TITLE
Cracking with multiple wordlists

### ISSUE THREAD

**Issue by alrai** [s3inlc/hashtopolis#567]
Hashtopolis installation info:
- Your current Server version located at the bottom of any Hashtopolis webpage: **0.11.0**
- Current Client version: **s3-python-0.5.0**
- Your current Hashcat version: **v5.1.0**
- The exact task command you are trying to run:
1) ./hashcat64.bin --keyspace --quiet ../../files/rockyou.txt ../../files/testlist.txt  --hash-type=0
2) ./hashcat64.bin --machine-readable --quiet --status --restore-disable --session=hashtopolis --status-timer 5 --outfile-check-timer=5 --outfile-check-dir=../../hashlist_7 -o ../../hashlists/7.out --outfile-format=15 -p "	" -s 0 -l 219098 --potfile-disable --remove --remove-timer=5 ../../hashlists/7 ../../files/rockyou.txt ../../files/testlist.txt  --hash-type=1000
- Debug output from the client by running "hashtopolis.exe -d" or with debug flag set on the python client:
1) Error during keyspace measure: Command './hashcat64.bin --keyspace --quiet ../../files/rockyou.txt ../../files/testlist.txt  --hash-type=0 ' returned non-zero exit status 255
2) Use of --skip/--limit is not supported with --increment or mask files.

I was trying to launch a task with two wordlists (selected both of them in the checkboxes as shown in the screenshot), but always encountered an error at the keyspace calculation stage (Error during keyspace measure: Command './hashcat64.bin --keyspace --quiet ../../files/rockyou.txt ../../files/testlist.txt  --hash-type=0 ' returned non-zero exit status 255).

![hashtopolis_task_creation](https://user-images.githubusercontent.com/22222912/60949902-cc084e00-a2fe-11e9-949a-c45d893e3658.png)

Obviously, the error occurred because hashcat can calculate a keyspace only for one wordlist at a time (when executed with --keyspace option). I re-wrote function **measure_keyspace** to make it iterate over wordlists calculating a keyspace for each and then a sum of all keyspaces:

```python
def measure_keyspace(self, task, chunk):
        if task['usePrince']:
            return self.prince_keyspace(task, chunk)
        keyspace2 = []
        for i in update_files(task['attackcmd']).replace(task['hashlistAlias'] + " ", "").split(' '):
            full_cmd = self.callPath + " --keyspace --quiet " + i + ' ' + task['cmdpars']
            if 'useBrain' in task and task['useBrain']:
                full_cmd += " -S"
            if Initialize.get_os() == 1:
                full_cmd = full_cmd.replace("/", '\\')
            try:
                logging.debug("CALL: " + full_cmd)
                output = subprocess.check_output(full_cmd, shell=True, cwd=self.cracker_path)
            except subprocess.CalledProcessError as e:
                logging.error("Error during keyspace measure: " + str(e))
                send_error("Keyspace measure failed!", self.config.get_value('token'), task['taskId'], None)
                sleep(5)
                return False
            output = output.decode(encoding='utf-8').replace("\r\n", "\n").split("\n")
            keyspace = "0"
            for line in output:
                if not line:
                    continue
                keyspace = line
            keyspace2.append(int(keyspace))
        return chunk.send_keyspace(sum(keyspace2), task['taskId'])
```

After this modification I was able to successfully pass the keyspace calculation stage, however another error emerged when the cracking started:

CALL: ./hashcat64.bin --machine-readable --quiet --status --restore-disable --session=hashtopolis --status-timer 5 --outfile-check-timer=5 --outfile-check-dir=../../hashlist_7 -o ../../hashlists/7.out --outfile-format=15 -p "	" -s 0 -l 219098 --potfile-disable --remove --remove-timer=5 ../../hashlists/7 ../../files/rockyou.txt ../../files/testlist.txt  --hash-type=1000
started cracking
HC error: Use of --skip/--limit is not supported with --increment or mask files

Please let me know if these errors can be fixed and if a task in general can use two or more wordlists.

Thank you!

**Comment by s3inlc** [comment]
As you realized correctly, hashcat cannot calculate the keyspace for more than wordlist. This is, because each wordlist itself is a task. And therefore you need to do two tasks for this. 
Of course your workaround would work for calculating the keyspace, but what if now a chunk contains a part of the first wordlist and a part of the second wordlist. This could not be run with a single hashcat command.
Hashcat internally never runs anything with two wordlists. When you provide it two wordlists, it will basically run two tasks, they are not mixed at all. That's why it also cannot calculate a keyspace, as it always looks at one wordlist only.

TL:DR; You just need to create two hashtopolis tasks, one with the first wordlist and one with the second. To make it easy, you can create the task with the first wordlist, and then just copy it and change the file.

**Comment by alrai** [comment]
Hi Sein,

Thanks for your responce.

As far as I know, [there is an ability to specify several dictionaries or even a directory containing several dictionaries on the command line for hashcat](https://hashcat.net/wiki/doku.php?id=frequently_asked_questions#how_to_use_multiple_dictionaries). It is only keyspace calculation which can't be performed on multiple files.

In practice we usually run hashcat with 50+ dictionaries, so creating a task for each dictionary would be extermely tedious. Can the source code be modified so that multiple tasks are created automatically if several dictionaries are selected in GUI (or allow to run a task with several dictionaries)? Perhaps I myself could come up with a workaround which would meet our requirements if you gave me a hint on how to achieve this.

**Comment by hops** [comment]
What you could do is configure your 50+ dictionaris as preconfigured tasks (one task per dictionary) and then combine them in a supertask.
Once you have a supertask you can easily apply that to a hashlist.
@s3inlc correct me if I'm wrong :)

**Comment by evilmog** [comment]
Its under BULK Import, there's an iterator for this, you can also use the API to create multiple preconfigured tasks and then add all of those to the supertask which is probably easier. There's a python 2 and python 3 binding set for the user-api.

https://github.com/evilmog/htpclientapi

https://github.com/s3inlc/hashtopolis/tree/master/doc/user-api

**Comment by evilmog** [comment]
Its under
```
Tasks -> Import Super Task
```

Then
```
Wordlist/Rule Bulk
```

Under your task definition set something like this
```
#HL# -a 0 FILE -r rule.rule
```
select rule.rule as your base, and file as your iterator, you can also reverse this, it will create the preconfigured tasks and the supertask for you

**Comment by s3inlc** [comment]
Thanks @hops and @evilmog for the provided help. Indeed the bulk creation would be the easiest solution if you want to create the 50+ pretasks you would ususally run.

> As far as I know, there is an ability to specify several dictionaries or even a directory containing several dictionaries on the command line for hashcat. It is only keyspace calculation which can't be performed on multiple files.

As I wrote in the first response, yes it is possible, but internally in hashcat this is handled as it would run all these dictionaries as separate tasks. And therefore all the structures which hashcat provides to distribute the task (--keyspace/--skip/--limit) are not possible to be used with that as technically these are 50+ tasks then.

**Comment by winxp5421** [comment]
I believe this was answered and a solution was found.
If you feel this issue was closed in error feel free to re-open and update us.
